File: server/internal/api/group_messages.go
// Сообщения групп (api-overview.md §Группы (переписка)): POST/GET
// /groups/{id}/messages. Private-группа: payload = SecretBox(zstd(MessageBody), GK),
// gk_version обязателен; публичная: plaintext protobuf, gk_version отсутствует.
// Подпись Ed25519 устройства по group_message_canonical_bytes
// (schema/proto/README.md; KAT group_message_canonical) проверяется при приёме.
//
// Премодерация (pending) — когда data-model §4 получит колонку статуса;
// сообщения от имени сущности (sender_type='entity', боты) — вместе с Bot API.
package api

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"log"
	"net/http"
	"strconv"
	"time"

	"tima/server/internal/auth"
	timacrypto "tima/server/internal/crypto"
	"tima/server/internal/escrow"
	"tima/server/internal/store"
)

// postGroupMessage — POST /groups/{groupID}/messages.
func postGroupMessage(deps groupsDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		g, role, ok := groupAndRole(deps, w, r)
		if !ok {
			return
		}
		if role == "" {
			writeErr(w, http.StatusNotFound, "group_not_found", "группа не найдена")
			return
		}
		groupID := r.PathValue("groupID")
		id, _ := auth.FromContext(r.Context())

		// Бан и slow mode (модераторам и выше slow mode не действует)
		_, bannedUntil, err := deps.store.GroupMemberInfo(r.Context(), groupID, id.UserID)
		if err != nil {
			log.Printf("postGroupMessage: member info: %v", err)
			writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
			return
		}
		if bannedUntil != nil && bannedUntil.After(time.Now()) {
			writeErr(w, http.StatusForbidden, "banned", "участник заблокирован до "+bannedUntil.UTC().Format(time.RFC3339))
			return
		}
		if g.SlowModeSec > 0 && roleRank[role] < rankModerator {
			recent, err := deps.store.SenderPostedWithin(r.Context(), groupID, id.UserID, g.SlowModeSec)
			if err != nil {
				log.Printf("postGroupMessage: slow mode: %v", err)
				writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
				return
			}
			if recent {
				w.Header().Set("Retry-After", strconv.Itoa(int(g.SlowModeSec)))
				writeErr(w, http.StatusTooManyRequests, "slow_mode", "подождите slow mode")
				return
			}
		}

		var req struct {
			ClientMsgID     string `json:"client_msg_id"`
			Kind            uint32 `json:"kind"`
			Level           *int16 `json:"level"` // nil = умолчание по виду группы
			GKVersion       int32  `json:"gk_version"`
			Payload         string `json:"payload"` // base64url
			ThreadRoot      int64  `json:"thread_root"`
			ReplyTo         int64  `json:"reply_to"`
			CreatedAtUnixMs int64  `json:"created_at_unix_ms"`
			Signature       string `json:"signature"` // base64url, Ed25519
		}
		if err := json.NewDecoder(io.LimitReader(r.Body, maxEnvelopeBytes+64<<10)).Decode(&req); err != nil {
			writeErr(w, http.StatusBadRequest, "bad_json", "тело не парсится")
			return
		}
		if req.ClientMsgID == "" {
			writeErr(w, http.StatusBadRequest, "no_client_msg_id", "нужен client_msg_id (UUID)")
			return
		}
		b64 := base64.RawURLEncoding
		payload, err1 := b64.DecodeString(req.Payload)
		signature, err2 := b64.DecodeString(req.Signature)
		if err1 != nil || err2 != nil || len(signature) != 64 {
			writeErr(w, http.StatusBadRequest, "bad_encoding", "payload/signature — base64url, подпись 64 байта")
			return
		}
		if len(payload) == 0 || len(payload) > maxEnvelopeBytes {
			writeErr(w, http.StatusBadRequest, "bad_payload", "payload пуст или больше 4 MiB")
			return
		}
		if req.ThreadRoot < 0 || req.ReplyTo < 0 || req.GKVersion < 0 {
			writeErr(w, http.StatusBadRequest, "bad_refs", "thread_root/reply_to/gk_version не могут быть отрицательными")
			return
		}

		// Уровень: кому сервер отдаст сообщение (ADR-0019). В подпись не входит —
		// он меняется после отправки, а подпись неизменна.
		level := defaultLevel(g.Kind)
		if req.Level != nil {
			level = *req.Level
		}
		if level < levelSecret || level > levelByGrant {
			writeErr(w, http.StatusBadRequest, "bad_level", "уровень вне диапазона -1…3")
			return
		}
		// Личная группа знает только -1 и 0 (ADR-0019 §2): всё, кроме описания,
		// зашифровано, а градаций внутри шифра нет.
		if g.Kind == "private" && level > levelPublicShowcase {
			writeErr(w, http.StatusBadRequest, "level_in_private", "в личной группе бывают только уровни -1 и 0")
			return
		}
		// В публичной группе шифра нет вовсе, значит и -1 неоткуда взяться.
		if g.Kind == "public" && level == levelSecret {
			writeErr(w, http.StatusBadRequest, "secret_in_public", "публичная группа не шифруется: уровень -1 недоступен")
			return
		}
		// Открытый текст на сервере ограничен по размеру: без предела он превращается
		// в бесплатный хостинг. Точный предел — 4096 знаков UTF-16 — проверяет клиент,
		// потому что знает содержимое; сервер видит байты и держит грубую границу.
		if level >= levelPublicShowcase && len(payload) > maxPlainPayloadBytes {
			writeErr(w, http.StatusBadRequest, "payload_too_large", "открытое сообщение больше 16 KiB")
			return
		}

		// Крипто-инварианты по типу группы
		if g.Kind == "private" && level == levelSecret {
			if req.GKVersion == 0 {
				writeErr(w, http.StatusBadRequest, "no_gk_version", "private-группа: нужен gk_version")
				return
			}
			if len(payload) < 24+16 {
				writeErr(w, http.StatusBadRequest, "bad_payload", "payload короче минимума SecretBox")
				return
			}
			known, err := deps.store.GroupKeyVersionExists(r.Context(), groupID, req.GKVersion)
			if err != nil {
				log.Printf("postGroupMessage: gk version: %v", err)
				writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
				return
			}
			if !known {
				writeErr(w, http.StatusBadRequest, "unknown_gk_version", "такой версии GK у группы нет")
				return
			}
		} else if req.GKVersion != 0 {
			// Незашифрованное сообщение с версией ключа — противоречие: версия говорит,
			// что payload закрыт, а уровень — что открыт.
			writeErr(w, http.StatusBadRequest, "gk_without_secret", "gk_version бывает только у уровня -1")
			return
		}

		// Ссылки веток/ответов — на сообщения этой же группы
		for name, ref := range map[string]int64{"thread_root": req.ThreadRoot, "reply_to": req.ReplyTo} {
			if ref == 0 {
				continue
			}
			exists, err := deps.store.GroupMessageExists(r.Context(), groupID, ref)
			if err != nil {
				log.Printf("postGroupMessage: ref %s: %v", name, err)
				writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
				return
			}
			if !exists {
				writeErr(w, http.StatusBadRequest, "bad_"+name, name+" не найден в группе")
				return
			}
		}

		// Подпись: устройство из токена, preimage — group_message_canonical_bytes
		signingPub, err := deps.store.SigningKey(r.Context(), id.DeviceID, id.UserID)
		if errors.Is(err, store.ErrDeviceUnknown) {
			writeErr(w, http.StatusForbidden, "unknown_device", "устройство отправителя не зарегистрировано")
			return
		} else if err != nil {
			log.Printf("postGroupMessage: signing key: %v", err)
			writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
			return
		}
		cb := timacrypto.GroupMessageCanonicalBytes(timacrypto.GroupMessageMeta{
			GroupID:         groupID,
			SenderID:        id.UserID,
			SenderDevice:    id.DeviceID,
			Kind:            req.Kind,
			CreatedAtUnixMs: req.CreatedAtUnixMs,
			ThreadRoot:      uint64(req.ThreadRoot),
			ReplyTo:         uint64(req.ReplyTo),
			GKVersion:       uint32(req.GKVersion),
		}, payload)
		if !timacrypto.VerifyEnvelopeSignature(signingPub, cb, signature) {
			writeErr(w, http.StatusForbidden, "bad_signature", "подпись сообщения не прошла проверку")
			return
		}

		msg := store.GroupMessage{
			GroupID:         groupID,
			ClientMsgID:     req.ClientMsgID,
			SenderID:        id.UserID,
			SenderDevice:    id.DeviceID,
			Kind:            int32(req.Kind),
			GKVersion:       req.GKVersion,
			Payload:         payload,
			ThreadRoot:      req.ThreadRoot,
			ReplyTo:         req.ReplyTo,
			CreatedAtUnixMs: req.CreatedAtUnixMs,
			Signature:       signature,
			Level:           level,
		}
		messageID, duplicate, err := deps.store.SaveGroupMessage(r.Context(), msg)
		if err != nil {
			log.Printf("postGroupMessage: save: %v", err)
			writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		if duplicate {
			_ = json.NewEncoder(w).Encode(map[string]any{"duplicate": true, "message_id": messageID})
			return
		}
		msg.MessageID = messageID

		// Доставка активным участникам (кроме устройства отправителя):
		// event log + live онлайн-устройствам
		devices, err := deps.store.ActiveMemberDevices(r.Context(), groupID, id.DeviceID)
		if err != nil {
			log.Printf("postGroupMessage: member devices: %v", err)
		}
		eventPayload := groupMessageJSON(msg)
		for _, dev := range devices {
			deps.notifier.Device(r.Context(), dev, "message.group", eventPayload)
		}

		// Инвариант ADR-0017 §2: в группе, где в эту эпоху была отправка, последняя версия
		// GK обязана быть завёрнута на текущую эпоху. Отправка только что произошла — самое
		// время проверить. Сервер ротировать не может (ключа он не видит), поэтому просит
		// участников: кто первым откроет приложение, тот и сменит.
		//
		// ТОЛЬКО ДЛЯ ЗАШИФРОВАННОГО. Инвариант говорит про отправку, которая пользовалась
		// ключом: escrow-блоб у группы один на версию GK, и восстанавливать по ордеру
		// нужно то, что закрыто. Открытое сообщение (уровни 0…3) ключа не касается и
		// лежит на сервере открытым — восстанавливать в нём нечего.
		//
		// Без этого условия описание личной группы — сообщение уровня 0 — гнало бы тихую
		// группу на ротацию, то есть на фан-аут обёрток по всем устройствам всех
		// участников. Недосмотр этапа Г1, где открытые сообщения появились; исправлено
		// 2026-09-04.
		if level == levelSecret {
			remindAboutRotation(deps, r.Context(), groupID, append(devices, id.DeviceID))
		}

		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(map[string]any{"message_id": messageID})
	}
}

// groupMessageJSON — сообщение в JSON-обвязке (все поля preimage — получатель
// обязан суметь проверить подпись).
func groupMessageJSON(m store.GroupMessage) map[string]any {
	b64 := base64.RawURLEncoding
	return map[string]any{
		"message_id":         m.MessageID,
		"group_id":           m.GroupID,
		"sender_id":          m.SenderID,
		"sender_device":      m.SenderDevice,
		"kind":               m.Kind,
		"gk_version":         m.GKVersion,
		"payload":            b64.EncodeToString(m.Payload),
		"thread_root":        m.ThreadRoot,
		"reply_to":           m.ReplyTo,
		"created_at_unix_ms": m.CreatedAtUnixMs,
		"signature":          b64.EncodeToString(m.Signature),
		"level":              m.Level,
	}
}

// listGroupMessages — GET /groups/{groupID}/messages?before=&limit=&thread=.
func listGroupMessages(deps groupsDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		g, role, ok := groupAndRole(deps, w, r)
		if !ok {
			return
		}
		// Не-участник получает историю только до уровня «всем» — и только если карточка
		// ему открыта. Иначе 404: существование личной группы не должно быть видно.
		if role == "" {
			id, _ := auth.FromContext(r.Context())
			open, err := cardOpenTo(deps, r, g.Kind, g.GroupID, id.UserID)
			if err != nil {
				log.Printf("listGroupMessages: card audience: %v", err)
				writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
				return
			}
			if !open {
				writeErr(w, http.StatusNotFound, "group_not_found", "группа не найдена")
				return
			}
		}
		var before, thread int64
		if v := r.URL.Query().Get("before"); v != "" {
			before, _ = strconv.ParseInt(v, 10, 64)
		}
		if v := r.URL.Query().Get("thread"); v != "" {
			thread, _ = strconv.ParseInt(v, 10, 64)
		}
		limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))

		// Граница выдачи: роль плюс поимённое разрешение, если оно есть и не истекло.
		max := maxLevelFor(role)
		if role != "" && max < levelByGrant {
			id, _ := auth.FromContext(r.Context())
			if granted, err := deps.store.GrantedLevelFor(r.Context(), r.PathValue("groupID"), id.UserID); err != nil {
				log.Printf("listGroupMessages: grant: %v", err)
			} else if granted > max {
				max = granted
			}
		}
		msgs, err := deps.store.ListGroupMessages(r.Context(), r.PathValue("groupID"), thread, before, limit, max)
		if err != nil {
			log.Printf("listGroupMessages: %v", err)
			writeErr(w, http.StatusInternalServerError, "internal", "ошибка хранилища")
			return
		}
		out := make([]map[string]any, 0, len(msgs))
		for _, m := range msgs {
			out = append(out, groupMessageJSON(m))
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"messages": out})
	}
}

// remindAboutRotation рассылает group.rotation_needed, если ключ группы завёрнут на
// прошлую эпоху (ADR-0017 §3).
//
// **Не чаще раза в час на группу и эпоху.** Иначе каждое сообщение в группе с устаревшим
// ключом порождало бы фан-аут события на все устройства — то есть наказывало бы за
// разговорчивость. Одного напоминания достаточно: получивший его клиент ротирует, и
// условие перестанет выполняться само.
//
// Без Redis (dev) напоминание уходит на каждое сообщение: это шумно, но честнее, чем
// молчать о невыполнимом ордере.
func remindAboutRotation(deps groupsDeps, ctx context.Context, groupID string, devices []string) {
	latest, err := deps.store.LatestGroupRotation(ctx, groupID)
	if err != nil {
		log.Printf("напомнитьОРотации: последняя ротация: %v", err)
		return
	}
	// Группа без ротаций ключа не имеет вовсе — напоминать не о чем: первую версию
	// выпустят при первой отправке в private-группу.
	if latest.GKVersion == 0 {
		return
	}
	epoch := escrow.EpochOf(time.Now())
	if latest.EscrowEpoch == epoch {
		return
	}
	// Эпоха сменилась — значит ключ всё равно будет ротироваться. Ровно здесь выводим
	// тех, чей срок участия вышел (ADR-0019 §9): выбытие обязано менять ключ, и делать
	// его в момент, когда ключ и так меняется, дешевле всего. Отдельной задачи по
	// расписанию поэтому не заводится — это было условием решения о сроках.
	//
	// Читать группу просроченный перестаёт раньше и без этого: `GroupRole` проверяет
	// срок тем же запросом, которым читает роль. Здесь — только уборка состава.
	if n, err := deps.store.ExpireMemberships(ctx, groupID); err != nil {
		log.Printf("expire memberships %s: %v", groupID, err)
	} else if n > 0 {
		log.Printf("группа %s: срок вышел у %d участников, ключ сменится ротацией эпохи", groupID, n)
	}
	if deps.limiter() != nil {
		ok, _, err := deps.limiter().Allow(ctx, "gk_epoch_notify:"+groupID+":"+epoch, 1, time.Hour)
		if err != nil || !ok {
			return
		}
	}
	for _, dev := range devices {
		deps.notifier.Device(ctx, dev, "group.rotation_needed", map[string]any{
			"group_id": groupID,
			"reason":   "epoch",
			"epoch":    epoch,
		})
	}
}

File: client/core/core-encryption/src/commonMain/kotlin/io/tima/core/encryption/GroupMessages.kt
package io.tima.core.encryption

import io.tima.crypto.CanonicalBytes
import io.tima.crypto.EnvelopeCipher
import io.tima.crypto.GroupMessageMeta
import io.tima.crypto.MessageContent
import io.tima.crypto.MessageContentCodec
import io.tima.crypto.MessageSerializer
import io.tima.crypto.MessageSigner
import io.tima.crypto.VerificationFailure

/**
 * Сообщение группы: собрать и открыть — один вызов вместо четырёх.
 *
 * **Чем это отличается от личного сообщения — не «почти тем же самым».** Личное едет
 * конвертом `Envelope`: у него на каждое сообщение свой одноразовый ключ, свои обёртки на
 * устройства получателей и свой escrow-блоб. У группового нет ничего из этого
 * (`crypto-protocol §4.1`): ключ один на версию GK, обёртки раздаются при ротации, escrow
 * тоже один на версию. Поэтому на провод уходит не конверт, а три поля — `payload`,
 * `signature` и метаданные, — и подпись считается по своей раскладке
 * (`group_message_canonical_bytes`), с доменной меткой, не пересекающейся с конвертом.
 *
 * Порядок шагов нормативен: содержимое → тело (`zstd(protobuf)`) → шифрование ключом
 * группы → подпись по каноническим байтам. Перепутанный порядок даёт не ошибку сборки,
 * а сообщение, которое не откроется ни у кого.
 *
 * ── ПОЧЕМУ ПУБЛИЧНАЯ ГРУППА СЮДА НЕ ПРОХОДИТ ─────────────────────────────────
 *
 * В протоколе `gk_version = 0` означает публичную группу, где `payload` идёт открытым
 * protobuf. Соблазнительно разрешить это здесь же: «нет ключа — шлём открытым». Именно так
 * и выглядит утечка: группа считалась частной, ключ не доехал, сообщение ушло в
 * незашифрованном виде, и никто ничего не заметил, потому что оно доставилось.
 *
 * Поэтому этот фасад требует версию ключа не меньше единицы. Публичная группа, когда она
 * появится, получит **отдельный** вызов, и его придётся написать намеренно.
 */
object GroupMessages {

    /**
     * @param groupKey GK той версии, что названа в [GroupMessageMeta.gkVersion]. Брать
     *   последнюю известную: сообщение, зашифрованное старой версией, участники, пришедшие
     *   после ротации, не прочитают.
     */
    fun seal(
        content: MessageContent,
        meta: GroupMessageMeta,
        sender: DeviceIdentity,
        groupKey: ByteArray,
    ): Result<SealedGroupMessage> = runCatching {
        require(meta.gkVersion >= 1) {
            "версия группового ключа обязана быть не меньше 1: нулевая означает открытый payload"
        }
        val body = MessageContentCodec.toBody(content)
        val plaintext = MessageSerializer.encodeBody(body)
        val payload = EnvelopeCipher.seal(groupKey, plaintext).getOrThrow()
        val canonical = CanonicalBytes.buildGroupMessage(meta, payload)
        SealedGroupMessage(
            meta = meta,
            payload = payload,
            signature = MessageSigner.sign(sender.key, canonical).getOrThrow(),
        )
    }

    /**
     * Собирает ОТКРЫТОЕ сообщение группы — уровни 0…3 (ADR-0019 §1).
     *
     * Шифрования здесь нет по построению: сообщение уровня 0 существует ровно для того,
     * чтобы его прочёл тот, у кого ключа нет и не будет, — карточка личной группы, лента
     * публичной. Подпись при этом остаётся: открытость содержимого не отменяет вопроса,
     * кто его написал.
     *
     * `gkVersion` обязан быть нулевым: ненулевой означал бы, что payload закрыт, и сервер
     * такое сообщение отвергнет (`gk_without_secret`).
     */
    fun sealPlain(
        content: MessageContent,
        meta: GroupMessageMeta,
        sender: DeviceIdentity,
    ): Result<SealedGroupMessage> = runCatching {
        require(meta.gkVersion == 0) {
            "у открытого сообщения версии ключа нет: ненулевая означает шифр"
        }
        val payload = MessageSerializer.encodeBody(MessageContentCodec.toBody(content))
        val canonical = CanonicalBytes.buildGroupMessage(meta, payload)
        SealedGroupMessage(
            meta = meta,
            payload = payload,
            signature = MessageSigner.sign(sender.key, canonical).getOrThrow(),
        )
    }

    /**
     * Открывает пришедшее сообщение группы.
     *
     * Подпись проверяется **до** расшифровки и по тем же метаданным, что пришли с сервером:
     * иначе сервер мог бы подменить автора или время, оставив payload нетронутым, и подпись
     * всё равно сошлась бы — она считается по метаданным вместе с содержимым.
     *
     * Провал возвращается как `Result.failure`, и рода у него два, различать их обязан
     * вызывающий: [VerificationFailure] — подпись не сошлась, то есть возможная подмена;
     * остальное — не тот ключ или повреждённые байты, то есть своя несобранная картина.
     */
    fun open(
        sealed: SealedGroupMessage,
        senderSigningPublic: ByteArray,
        groupKey: ByteArray,
    ): Result<ReceivedGroupMessage> = runCatching {
        val canonical = CanonicalBytes.buildGroupMessage(sealed.meta, sealed.payload)
        if (!MessageSigner.verify(senderSigningPublic, canonical, sealed.signature)) {
            throw VerificationFailure("Подпись сообщения группы не прошла проверку")
        }
        val plaintext = EnvelopeCipher.open(groupKey, sealed.payload).getOrThrow()
        val body = MessageSerializer.decodeBody(plaintext).getOrThrow()
        ReceivedGroupMessage(
            meta = sealed.meta,
            content = MessageContentCodec.fromBody(body),
            // Байты тела отдаются как пришли — уже упакованными. Хранилище пишет именно их:
            // кодек один на провод и на диск. Записать текстом значит записать в другом
            // формате, чем читает экран, — на этом уже ловились личные сообщения.
            body = plaintext,
        )
    }

    /**
     * То же открытие, но **по полям кадра**, как они пришли с сервера.
     *
     * Появилось 2026-08-25. До этого приёмник собирал [GroupMessageMeta] сам — восемь
     * полей в правильном порядке и с правильными типами, — и ради этого импортировал
     * крипто-типы в composition root. Плата была не в строчках: раскладка метаданных
     * входит в подписываемые байты, значит любое её изменение обязано было править и
     * приёмник, а забытая правка даёт не ошибку сборки, а сообщение, которое не
     * открывается ни у кого.
     *
     * Теперь meta собирается здесь, рядом с `CanonicalBytes`, — в том же модуле, что
     * и правило её раскладки.
     *
     * @param threadRoot и [replyTo] — беззнаковые в протоколе, но приходят из транспорта
     *   знаковыми `Long`: преобразование сделано здесь, чтобы вызывающий о нём не знал.
     */
    @Suppress("LongParameterList")
    fun open(
        groupId: String,
        senderId: String,
        senderDevice: String,
        kind: Int,
        createdAtUnixMs: Long,
        threadRoot: Long,
        replyTo: Long,
        gkVersion: Int,
        payload: ByteArray,
        signature: ByteArray,
        senderSigningPublic: ByteArray,
        groupKey: ByteArray,
    ): Result<ReceivedGroupMessage> = open(
        sealed = SealedGroupMessage(
            meta = GroupMessageMeta(
                groupId = groupId,
                senderId = senderId,
                senderDevice = senderDevice,
                kind = kind,
                createdAtUnixMs = createdAtUnixMs,
                threadRoot = threadRoot.toULong(),
                replyTo = replyTo.toULong(),
                gkVersion = gkVersion,
            ),
            payload = payload,
            signature = signature,
        ),
        senderSigningPublic = senderSigningPublic,
        groupKey = groupKey,
    )
}

/** Готовое к отправке сообщение группы: то, что уходит в `POST /groups/{id}/messages`. */
class SealedGroupMessage(
    val meta: GroupMessageMeta,
    /** `SecretBox(zstd(protobuf(body)), GK)`. */
    val payload: ByteArray,
    /** Ed25519 по `group_message_canonical_bytes`, 64 байта. */
    val signature: ByteArray,
)

/** Разобранное сообщение группы. [body] — байты тела как пришли, их и пишет хранилище. */
class ReceivedGroupMessage(
    val meta: GroupMessageMeta,
    val content: MessageContent,
    val body: ByteArray,
)
